### PR TITLE
decider: endtoend test infrastructure + tests

### LINKED ISSUE #6769: Task: vtorc endtoend tests
Now that we have a vitess-native implementation of [orchestrator](https://github.com/openark/orchestrator) aka vtorc, we need to add endtoend tests for the following use cases.
- [x] Cluster has no master, master should be elected automatically
- [x] Cluster has no master, RDONLY should not be elected as master
- [x] Bring master down, a replica should be elected, replication from other instances should be re-configured to point to new master
- [ ] Bring up old master, should rejoin cluster properly
- [x] Graceful takeover
- [x] TER to a non-master, wait for repair
- [x] Make Master read-only, wait for repair
- [x] stop replication on a replica, wait for repair
- [x] point replica elsewhere, wait for repair
- [x] make replica read-write, wait for repair
- [x] Create a co-master situation (instance A replicates from B and B from A), wait for repair
- [x] Basic testcases for `-clusters_to_watch` (default, single keyspace, single keyspace/shard)
- [x] In a multi-cluster environment, each orc instance acts on specified clusters (`-clusters_to_watch` flag)

Infrastructure tasks:
- [x] basic test infrastructure
- [x] enable in CI by adding to `test/config.json`
- [x] configuration changes to speed up tests
- [x] generate port number instead of using default

EDIT June 1, 2021:
Replace Decider with vtorc.

### PR DESCRIPTION
**Pull request by deepthi**
Partial fix for #6769 

Signed-off-by: deepthi <deepthi@planetscale.com>

### CODE REVIEW

File: go/test/endtoend/orchestrator/orc_test.go
Comment (sougou) [inline go/test/endtoend/orchestrator/orc_test.go:145]: This shouldn't be needed any more. Starting the vttablet should create the db automatically. Also, I'm surprised this worked because the replicas will all have diverging gtids.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (sougou) [inline go/test/endtoend/orchestrator/orc_test.go:219]: Better to keep vttablet running. Killing vttablet will be a different use case (and test), which orc doesn't handle yet.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (shlomi-noach) [inline go/test/endtoend/orchestrator/orc_test.go:224]: I'm just confused: where's the code that waits?

File: go/test/endtoend/orchestrator/orc_test.go
Comment (shlomi-noach) [inline go/test/endtoend/orchestrator/orc_test.go:242]: what does this function test?

File: go/test/endtoend/orchestrator/orc_test.go
Comment (shlomi-noach) [inline go/test/endtoend/orchestrator/orc_test.go:228]: ah, so this is a busy loop that waits for the tablet to get into some status? I think I understand now. I'd rather:

- avoid busy loop, insert some reasonable `time.Sleep(time.Second)` in the loop, or else these tests will be difficult to run on one's computer and in CI
- set some timeout for failure?
- or, really change the approach. orchestrator's system tests work the other way around: we setup the topology, we wait 15 seconds (by way of example) and then we expect action to have taken place.
  This approach makes sense for production, because in production, you not only expect orchestrator to run a failover, you also expect it to accomplish it within a reasonable timeframe. 
  Also, in this test, you wait until some state is known, and then immediately test that state. But there might be intermediate steps to the recovery that may promote one server, then another. So testing "as soon as there's some state" may land you in an intermediate situation.
IMHO waiting X seconds (as opposed to looping for some state) is a more correct approach, I don't feel strongly; if you add Sleep+timeout instead, that also can makes sense in some situations.

**Review by shlomi-noach — COMMENTED** [review]
Good work! Please see inline comment for `TestDownMaster()`, which is a general observation on the possible approaches to run the tests, and why one may be more advantageous than the other.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:145]: I'm not sure why this is _NOT_ creating problems, but I will remove the call to CreateDB.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:219]: I don't see that it makes a difference to this case because we are also killing MySQL. I agree that killing just tablet, not MySQL is a different case.

File: go/test/endtoend/cluster/cluster_process.go
Comment (deepthi) [reply go/test/endtoend/cluster/cluster_process.go:684]: All extraneous fields will be cleaned up before I move this PR from Draft. I kept that around in case I need it for one of the cases that hasn't been implemented yet.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:228]: It's not a busy loop. In fact, it implements something very similar to what you are suggesting - sleep 1 second between checks and timeout after 60 seconds.
I had initially tried a 15 second sleep right after starting orchestrator and checking for master tablet once. According to my testing, it takes ~21 seconds on my local machine to get a new master elected when I have `"RecoveryPeriodBlockSeconds": 1,` and ~41 seconds when I set that to 5.
Waiting for a fixed amount of time tends to be unreliable on CI so it is actually better to poll for a desired condition.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:242]: it doesn't _test_ anything. It is a utility function to get the current master tablet from topo.
The reason it loops and waits is because we could be waiting for a master to be elected.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:224]: later :) I'll remove the comment.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:228]: you make a good point here:
```
But there might be intermediate steps to the recovery that may promote one server, then another. So testing "as soon as there's some state" may land you in an intermediate situation.
```
Let us see if we do end up in that state in any of the scenarios we want to test and if we do we can improve the checks to be aware of the possibility.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (sougou) [reply go/test/endtoend/orchestrator/orc_test.go:185]: There is a subtle difference: if the vttablet shuts down, it will remove its port mappings from the tablet record. If the orchestrator loads the record after that, it will notice this change and remove the instance from the cluster.In that case, we'll end up in a "no current master" scenario.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (shlomi-noach) [reply go/test/endtoend/orchestrator/orc_test.go]: 41 seconds are far too long. I suspect MySQL configuration. Are replicas configured with: 

- `set global slave_net_timeout = 4`, see documentation. This sets a short (2sec) heartbeat interval between a replica and its master, and will make the replica recognize failure quickly. Without this setting, some scenarios may take up to a minute to detect.
- `CHANGE MASTER TO MASTER_CONNECT_RETRY=1, MASTER_RETRY_COUNT=86400`. In the event of replication failure, make the replica attempt reconnection every 1sec (default is 60sec). With brief network issues this setting attempts a quick replication recovery and, if successful, will avoid a general failure/recovery operation by orchestrator.


See https://github.com/openark/orchestrator/blob/master/docs/configuration-failure-detection.md#mysql-configuration

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go:185]: makes sense. I will fix it.

File: go/test/endtoend/orchestrator/orc_test.go
Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go]: I have changed the config file used by tests (`default-fast.cnf`) to set the short heartbeat interval.
I see that `MASTER_RETRY_COUNT` defaults to `86400` anyway. Regarding `MASTER_CONNECT_RETRY`, we are bringing up clusters with no master (and no replication setup) and allowing orc to elect the master. I think this means that we should make a change here to add `MASTER_CONNECT_RETRY=1`
https://github.com/vitessio/vitess/blob/master/go/vt/orchestrator/inst/instance_topology_dao.go#L677

Comment (deepthi) [reply go/test/endtoend/orchestrator/orc_test.go]: On second thoughts, the changes to `MASTER_RETRY_COUNT` can be in a separate PR (since we will have more PRs with tests in them). I have added a task to the issue to keep track.

**Review by sougou — APPROVED** [review]

**Pull request merged**
